spec-tests/src/test/java/org/candlepin/spec/pools/PoolSpecTest.java

```diff
@@ -0,0 +1,625 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.candlepin.ApiException;
+import org.candlepin.dto.api.v1.AttributeDTO;
+import org.candlepin.dto.api.v1.BrandingDTO;
+import org.candlepin.dto.api.v1.ConsumerDTO;
+import org.candlepin.dto.api.v1.ConsumerTypeDTO;
+import org.candlepin.dto.api.v1.EntitlementDTO;
+import org.candlepin.dto.api.v1.OwnerDTO;
+import org.candlepin.dto.api.v1.PoolDTO;
+import org.candlepin.dto.api.v1.ProductDTO;
+import org.candlepin.dto.api.v1.UserDTO;
+import org.candlepin.spec.bootstrap.client.ApiClient;
+import org.candlepin.spec.bootstrap.client.ApiClients;
+import org.candlepin.spec.bootstrap.client.SpecTest;
+import org.candlepin.spec.bootstrap.data.builder.Branding;
+import org.candlepin.spec.bootstrap.data.builder.Owners;
+import org.candlepin.spec.bootstrap.data.builder.Pools;
+import org.candlepin.spec.bootstrap.data.builder.Products;
+import org.candlepin.spec.bootstrap.data.util.StringUtil;
+import org.candlepin.spec.bootstrap.data.util.UserUtil;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.time.OffsetDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+
+
+@SpecTest
+class PoolSpecTest {
+
+    private OwnerDTO createOwner(ApiClient client) throws ApiException {
+        return client.owners().createOwner(Owners.random());
+    }
+
+    private ConsumerDTO createConsumer(ApiClient client, OwnerDTO owner, String name,
+        Map<String, String> facts) throws ApiException {
+
+        ConsumerTypeDTO ctype = new ConsumerTypeDTO()
+            .label("system");
+
+        ConsumerDTO consumer = new ConsumerDTO()
+            .name(name)
+            .owner(Owners.toNested(owner))
+            .type(ctype);
+
+        if (facts != null) {
+            // Copy the map into a new hashmap to ensure we can still modify it later if we so choose
+            consumer.facts(new HashMap<>(facts));
+        }
+
+        consumer.putFactsItem("system.certificate_version", "3.3");
+
+        return client.consumers()
+            .register(consumer);
+    }
+
+    private ProductDTO createProduct(ApiClient client, OwnerDTO owner) throws ApiException {
+        ProductDTO product = Products.randomSKU();
+
+        return client.ownerProducts()
+            .createProductByOwner(owner.getKey(), product);
+    }
+
+    // ugh...
+    private AttributeDTO buildAttribute(String key, String value) {
+        return new AttributeDTO()
+            .name(key)
+            .value(value);
+    }
+
+    private PoolDTO createPool(ApiClient client, OwnerDTO owner) throws ApiException {
+        ProductDTO product = this.createProduct(client, owner);
+        return this.createPool(client, owner, product);
+    }
+
+    private PoolDTO createPool(ApiClient client, OwnerDTO owner, ProductDTO product) throws ApiException {
+        PoolDTO pool = Pools.randomUpstream(product);
```

Comment:
_If_ we decide to extract these methods to utility classes for reusability, then this method (and the one above) are hiding the fact that they set the upstream subscription properties; so we should probably use a different method name for them to differentiate them from custom pools that don't have those properties.

Alternatively, since for the majority of use cases the pool being custom vs imported from upstream does not matter, we shouldn't be calling the randomUpstream method within any such utility method, and instead let the developer explicitly use it only in the few tests that need them

Reply:
If my memory is correct, the upstream requirement is more common than not. It ended up being a real pain when I tried to fully remove the ruby `create_subscription_and_pool` stuff some years back.

But more to your point, I don't think it's necessary to turn every local abstraction into a global utility method. Something I hit when attempting to do so with the consumer stuff, which is even more generic than this, is that turning this local method into a global function means more overloads and updating of existing functions to try to make every caller from every class able to use it. It ended up making them ugly and cumbersome, with no fewer than 4 variants of the signature.

That's kind of why the `randomUpstream` function exists -- I was requested to move a local context to a global one, but had to avoid breaking existing tests which I have no interest in updating in this PR. It doesn't add a ton of value IMO, and hides details about the pool creation that may or may not be critical to the individual test. And on the topic of the properties that make a pool appear like an upstream pool, rather than discouraging the use of such a function, I would much rather we sort out why those attributes are critical, and update pool so they're not. That's non-trivial, especially in the import bits where it's absolutely required today (it's how local pools are mapped to stuff from the manifest); but it'd be a much better long-term solution than trying to use function A and then falling back to function B if it doesn't work, without much thought to the details.